File: 0_1.baekjoon/implementation/python/BOJ21277.py
from sys import stdin
readline = stdin.readline

N1, M1 = map(int, readline().split())
mat1 = [list(map(int, list(readline().rstrip()))) for _ in range(N1)]
N2, M2 = map(int, readline().split())
mat2 = [list(map(int, list(readline().rstrip()))) for _ in range(N2)]

def solve():
    global mat2
    max_row = max(N1, N2)
    max_col = max(M1, M2)
    ans = float('inf')
    for _ in range(4):

        for row_offset in range(-max_row, max_row + 1):
            for col_offset in range(-max_col, max_col + 1):

                if not check(row_offset, col_offset): continue

                left = min(0, row_offset)
                right = max(N1, len(mat2) + row_offset)
                row_len = abs(left) + right

                left = min(0, col_offset)
                right = max(M1, len(mat2[0]) + col_offset)
                col_len = abs(left) + right

                ans = min(ans, row_len * col_len)

        mat2 = rotate(mat2)

    print(ans)

def check(row_offset, col_offset):

    for r in range(len(mat2)):
        for c in range(len(mat2[0])):
            nr, nc = r + row_offset, c + col_offset

            if nr < 0 or nr >= N1 or nc < 0 or nc >= M1: continue

            if mat1[nr][nc] == 1 and mat2[r][c] == 1:
                return False

    return True

def rotate(matrix):
    temp = [[0 for _ in range(len(matrix))] for _ in range(len(matrix[0]))]

    for r in range(len(matrix)):
        for c in range(len(matrix[0])):
            temp[c][len(matrix) - r - 1] = matrix[r][c]

    return temp
solve()

File: 0_1.baekjoon/implementation/java/src/BOJ21277.java
import java.io.*;
import java.util.*;

class BOJ21277 {
    static int N1, N2, M1, M2;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N1 = Integer.parseInt(st.nextToken());
        M1 = Integer.parseInt(st.nextToken());

        int[][] p1 = new int[N1][M1];

        for (int i=0;i<N1;i++) {
            String[] line = br.readLine().strip().split("");
            for (int j=0;j<M1;j++) {
                p1[i][j] = Integer.parseInt(line[j]);
            }
        }

        st = new StringTokenizer(br.readLine());

        N2 = Integer.parseInt(st.nextToken());
        M2 = Integer.parseInt(st.nextToken());

        
        int[][] p2 = new int[N2][M2];

        for (int i=0;i<N2;i++) {
            String[] line = br.readLine().strip().split("");
            for (int j=0;j<M2;j++) {
                p2[i][j] = Integer.parseInt(line[j]);
            }
        }

        System.out.println(Math.min(solve(p1, p2), solve(p2, p1)));
    }

    private static int solve(int[][] p1, int[][] p2) {
        int result = Integer.MAX_VALUE;
        int n1 = p1.length, m1 = p1[0].length;
        int n2 = p2.length, m2 = p2[0].length;

        for (int l=0;l<4;l++) {
            for (int k=0;k<4;k++) {
                for (int rowOffset=0;rowOffset<=n1;rowOffset++) {
                    for(int colOffset=0;colOffset<=m1;colOffset++) {
                        
                        int current = check(rowOffset, colOffset, p1, p2);
    
                        result = Math.min(result, current);
                    }
                }
                p2 = rotate(p2);
                int temp = n2;
                n2 = m2;
                m2 = temp;
            }    
            p1 = rotate(p1);
            int temp = n1;
            n1 = m1;
            n1 = temp;
        }
        return result;
    }

    private static int check(int rowOffset, int colOffset, int[][] p1, int[][] p2) {
        int n1 = p1.length, m1 = p1[0].length;
        int n2 = p2.length, m2 = p2[0].length;

        for (int i=0;i<n2;i++) {
            for (int j=0;j<m2;j++) {
                int row = i + rowOffset;
                int col = j + colOffset;

                if (row >= n1 || col >= m1) continue;

                if (p1[row][col] == 1 && p2[i][j] == 1)
                    return Integer.MAX_VALUE;
            }
        }

        int colLen = Math.max(colOffset + m2, m1);
        int rowLen = Math.max(rowOffset + n2, n1);

        return colLen * rowLen;
    }

    public static int[][] rotate(int[][] puzzle) {
        int n = puzzle.length;
        int m = puzzle[0].length;

        int[][] temp = new int[m][n];

        for (int i=0;i<n;i++) {
            for (int j=0;j<m;j++) {
                temp[j][n-1-i] = puzzle[i][j];
            }
        }

        return temp;
    }
}